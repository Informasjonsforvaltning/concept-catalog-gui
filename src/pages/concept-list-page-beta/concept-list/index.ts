export { ConceptList } from './components/concept-list';
