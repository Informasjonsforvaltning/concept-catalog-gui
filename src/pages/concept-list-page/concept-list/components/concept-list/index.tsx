import React, { FC, useState } from 'react';

import { Concept } from '../../../../../types';
import { ConceptField, SortDirection } from '../../../../../types/enums';
import { sortConceptsByKey } from '../../../../../utils/sort';

import {
  findOriginalConcepts,
  findRevisionsOfConcept,
  hasConceptAnyRevisions
} from '../../utils/concepts';
import ListItem from '../concept-list-item';
import ConceptListHeader from '../concept-list-header';
import CollapseItem from '../concept-list-collapse-item';

interface Props {
  items: Concept[];
}

export const ConceptList: FC<Props> = ({ items }) => {
  const [sortField, setSortField] = useState<string>(ConceptField.MODIFY_TIME);
  const [sortDirection, setSortDirection] = useState<SortDirection>(
    SortDirection.ASC
  );

  const onSortField = (
    field: string,
    updatedSortDirection: SortDirection
  ): void => {
    setSortField(field);
    setSortDirection(updatedSortDirection);
  };

  return (
    <div className='d-flex flex-column flex-fill'>
      <ConceptListHeader
        sortField={sortField}
        sortDirection={sortDirection}
        onSortField={onSortField}
      />
      {items &&
        findOriginalConcepts(items)
          .sort(sortConceptsByKey(sortField, sortDirection))
          .map((concept, index) =>
            hasConceptAnyRevisions(concept.originaltBegrep, items) ? (
              <CollapseItem
                key={index}
                concepts={findRevisionsOfConcept(concept.originaltBegrep, items)
                  .sort(sortConceptsByKey(sortField, sortDirection))
                  .sort(
                    (
                      { erSistPublisert: a = false }: Concept,
                      { erSistPublisert: b = false }: Concept
                    ) => Number(b) - Number(a)
                  )}
              />
            ) : (
              <ListItem key={`${concept?.id}-${index}`} concept={concept} />
            )
          )}
    </div>
  );
};
