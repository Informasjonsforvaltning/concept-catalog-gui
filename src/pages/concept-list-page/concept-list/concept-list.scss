@import '../../../../node_modules/designsystemet/fdk-designsystem-bootstrap4/scss/colors.scss';

.fdk-list-header {
  background-color: $color3;
  color: $color0;
  margin-bottom: 0.25em;
  border-radius: 3px;
  padding: 0.5em 1em;

  .header-item {
    font-weight: 500;
  }
}
