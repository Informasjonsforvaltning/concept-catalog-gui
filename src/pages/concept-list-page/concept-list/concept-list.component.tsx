import React, { useState } from 'react';
import _ from 'lodash';

import { localization } from '../../../lib/localization';
import { ListItem as ListItemInterface } from '../../../domain/ListItem';
import { ListItem } from './list-item/list-item.component';
import { SortButtons } from '../../../components/sort-button/sort-button.component';
import './concept-list.scss';

interface Props {
  items: ListItemInterface[];
}

const renderListHeader = (sortField, sortDirection, onSortField): JSX.Element => (
  <div className="row fdk-list-header">
    <div className="col-3 d-flex align-items-center">
      <span className="header-item mr-1">{localization['preferredTerm']}</span>
      <SortButtons field="title" sortField={sortField} sortType={sortDirection} onSortField={onSortField} />
    </div>

    <div className="col-3 d-flex align-items-center">
      <span className="header-item mr-1">{localization['fieldOfStudy']}</span>
      <SortButtons field="theme" sortField={sortField} sortType={sortDirection} onSortField={onSortField} />
    </div>

    <div className="col-3 d-flex align-items-center">
      <span className="header-item mr-1">{localization['validity']}</span>
      <SortButtons field="valid" sortField={sortField} sortType={sortDirection} onSortField={onSortField} />
    </div>

    <div className="col-3 d-flex align-items-center">
      <span className="header-item mr-1">{localization['status']}</span>
      <SortButtons field="status" sortField={sortField} sortType={sortDirection} onSortField={onSortField} />
    </div>
  </div>
);

const renderListItems = (items, sortField, sortDirection): JSX.Element | null => {
  if (!items) {
    return null;
  }
  return _.orderBy(items, sortField, [sortDirection]).map(
    (item, index): JSX.Element => {
      return (
        <ListItem
          key={`${_.get(item, 'title')}-${index}`}
          col1={_.get(item, 'title')}
          col2={_.get(item, 'theme')}
          col3={_.get(item, 'valid')}
          status={_.get(item, 'status')}
          path="/tester"
        />
      );
    }
  );
};

export const ConceptList = ({ items }: Props): JSX.Element | null => {
  const [sortField, setSortField] = useState();
  const [sortDirection, setSortDirection] = useState<string>('asc');

  const onSortField = (field, sortDirection): void => {
    setSortField(field);
    setSortDirection(sortDirection);
  };

  return (
    <div className="d-flex flex-column flex-fill">
      {renderListHeader(sortField, sortDirection, onSortField)}
      {renderListItems(items, sortField, sortDirection)}
    </div>
  );
};
