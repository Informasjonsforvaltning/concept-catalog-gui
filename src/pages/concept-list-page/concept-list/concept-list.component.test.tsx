import React from 'react';
import _ from 'lodash';
import { shallow } from 'enzyme';
import { ConceptList } from './concept-list.component';
import conceptItems from '../../../mock/concepts.json';

test('should render ConceptList', (): void => {
  const wrapper = shallow(<ConceptList items={_.get(conceptItems, 'concepts')} />);
  expect(wrapper).toMatchSnapshot();
});
