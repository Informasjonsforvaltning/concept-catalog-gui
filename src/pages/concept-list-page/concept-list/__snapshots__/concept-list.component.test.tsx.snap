// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ConceptList 1`] = `
<div
  className="d-flex flex-column flex-fill"
>
  <div
    className="row fdk-list-header"
  >
    <div
      className="col-3 d-flex align-items-center"
    >
      <span
        className="header-item mr-1"
      >
        Anbefalt term
      </span>
      <Component
        field="prefLabel.nb"
        onSortField={[Function]}
        sortType="asc"
      />
    </div>
    <div
      className="col-3 d-flex align-items-center"
    >
      <span
        className="header-item mr-1"
      >
        Fagområde
      </span>
      <Component
        field="subject.nb"
        onSortField={[Function]}
        sortType="asc"
      />
    </div>
    <div
      className="col-3 d-flex align-items-center"
    >
      <span
        className="header-item mr-1"
      >
        Gyldighet
      </span>
      <Component
        field="valid"
        onSortField={[Function]}
        sortType="asc"
      />
    </div>
    <div
      className="col-3 d-flex align-items-center"
    >
      <span
        className="header-item mr-1"
      >
        Status
      </span>
      <Component
        field="status"
        onSortField={[Function]}
        sortType="asc"
      />
    </div>
  </div>
  <Component
    col1="avvik"
    col2="Kvalitetsstyring"
    col3=""
    key="https://fellesdatakatalog.brreg.no/api/concepts/75b8cede-953e-44ea-8b56-d7a1c62e4536-0"
    path="/75b8cede-953e-44ea-8b56-d7a1c62e4536"
    status="DRAFT"
  />
  <Component
    col1="arkivskaper"
    col2="Arkiv"
    col3=""
    key="https://fellesdatakatalog.brreg.no/api/concepts/8d641622-9bb9-4316-b2ab-b27e083f893f-1"
    path="/8d641622-9bb9-4316-b2ab-b27e083f893f"
    status="DRAFT"
  />
  <Component
    col1="dokumentnummer"
    col2="Tinglysing"
    col3=""
    key="https://fellesdatakatalog.brreg.no/api/concepts/18701034-85f7-4cb2-accf-5b4d285e5440-2"
    path="/18701034-85f7-4cb2-accf-5b4d285e5440"
    status="DRAFT"
  />
  <Component
    col1="produkttype"
    col2="Tilgjengeliggjøring"
    col3=""
    key="https://fellesdatakatalog.brreg.no/api/concepts/0c5d72fe-1c08-4e2f-9772-825f20921ab0-3"
    path="/0c5d72fe-1c08-4e2f-9772-825f20921ab0"
    status="DRAFT"
  />
  <Component
    col1="pantedokument"
    col2="Tinglysing"
    col3=""
    key="https://fellesdatakatalog.brreg.no/api/concepts/13dbd4a3-4aff-498c-abdd-61d8f7d6fc40-4"
    path="/13dbd4a3-4aff-498c-abdd-61d8f7d6fc40"
    status="DRAFT"
  />
  <Component
    col1="bruksenhetsnummer"
    col2="Adresse"
    col3=""
    key="https://fellesdatakatalog.brreg.no/api/concepts/d45424f6-deb1-4768-8bfe-864b71b25b0b-5"
    path="/d45424f6-deb1-4768-8bfe-864b71b25b0b"
    status="DRAFT"
  />
  <Component
    col1="nedkvittering"
    col2="Tinglysing"
    col3=""
    key="https://fellesdatakatalog.brreg.no/api/concepts/8906d79d-8814-4eab-bd4a-ee8cf8f67ad5-6"
    path="/8906d79d-8814-4eab-bd4a-ee8cf8f67ad5"
    status="DRAFT"
  />
  <Component
    col1="kontaktinformasjon"
    col2="Enhetsregisteret"
    col3=""
    key="https://fellesdatakatalog.brreg.no/api/concepts/de5b25ad-521f-4fbd-af51-7486129c6a6f-7"
    path="/de5b25ad-521f-4fbd-af51-7486129c6a6f"
    status="DRAFT"
  />
  <Component
    col1="fødselsår"
    col2="Person"
    col3=""
    key="https://fellesdatakatalog.brreg.no/api/concepts/95d61746-73c5-41f3-a72d-ce4b583609c1-8"
    path="/95d61746-73c5-41f3-a72d-ce4b583609c1"
    status="DRAFT"
  />
  <Component
    col1="avsender"
    col2="Arkiv"
    col3=""
    key="https://fellesdatakatalog.brreg.no/api/concepts/1c770979-34b0-439c-a7cb-adacb3619927-9"
    path="/1c770979-34b0-439c-a7cb-adacb3619927"
    status="DRAFT"
  />
</div>
`;
